File: packages/react-native/src/private/testing/fantom/specs/NativeFantom.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict
 * @format
 */

import type {
  RootTag,
  TurboModule,
} from '../../../../../Libraries/TurboModule/RCTExport';

import * as TurboModuleRegistry from '../../../../../Libraries/TurboModule/TurboModuleRegistry';

// match RenderFormatOptions.h
export type RenderFormatOptions = {
  includeRoot: boolean,
  includeLayoutMetrics: boolean,
};

// match RawEvent.h
export enum NativeEventCategory {
  /*
   * Start of a continuous event. To be used with touchStart.
   */
  ContinuousStart = 0,

  /*
   * End of a continuous event. To be used with touchEnd.
   */
  ContinuousEnd = 1,

  /*
   * Priority for this event will be determined from other events in the
   * queue. If it is triggered by continuous event, its priority will be
   * default. If it is not triggered by continuous event, its priority will be
   * discrete.
   */
  Unspecified = 2,

  /*
   * Forces discrete type for the event. Regardless if continuous event is
   * ongoing.
   */
  Discrete = 3,

  /*
   * Forces continuous type for the event. Regardless if continuous event
   * isn't ongoing.
   */
  Continuous = 4,

  /*
   * Priority for events that can be processed in idle times or in the
   * background.
   */
  Idle = 5,
}

export type ScrollOptions = {
  x: number,
  y: number,
  zoomScale?: number,
};

export type ImageResponse = {
  width: number,
  height: number,
  cacheStatus?: 'memory' | 'disk' | 'disk/memory',
  errorMessage?: string,
};

interface Spec extends TurboModule {
  startSurface: (
    viewportWidth: number,
    viewportHeight: number,
    devicePixelRatio: number,
    viewportOffsetX?: number,
    viewportOffsetY?: number,
  ) => RootTag;
  stopSurface: (surfaceId: RootTag) => void;
  enqueueNativeEvent: (
    shadowNode: unknown /* ShadowNode */,
    type: string,
    payload?: unknown,
    category?: NativeEventCategory,
    isUnique?: boolean,
  ) => void;
  enqueueScrollEvent: (
    shadowNode: unknown /* ShadowNode */,
    options: ScrollOptions,
  ) => void;
  enqueueModalSizeUpdate: (
    shadowNode: unknown /* ShadowNode */,
    height: number,
    width: number,
  ) => void;
  takeMountingManagerLogs: (surfaceId: RootTag) => Array<string>;
  getDirectManipulationProps: (
    shadowNode: unknown /* ShadowNode */,
  ) => $ReadOnly<{
    [string]: unknown,
  }>;
  getFabricUpdateProps: (shadowNode: unknown /* ShadowNode */) => $ReadOnly<{
    [string]: unknown,
  }>;
  flushMessageQueue: () => void;
  flushEventQueue: () => void;
  produceFramesForDuration: (miliseconds: number) => void;
  validateEmptyMessageQueue: () => void;
  getRenderedOutput: (
    surfaceId: RootTag,
    config: RenderFormatOptions,
  ) => string;
  reportTestSuiteResultsJSON: (results: string) => void;
  createShadowNodeReferenceCounter(
    shadowNode: unknown /* ShadowNode */,
  ): () => number;
  createShadowNodeRevisionGetter(
    shadowNode: unknown /* ShadowNode */,
  ): () => ?number;
  saveJSMemoryHeapSnapshot: (filePath: string) => void;
  forceHighResTimeStamp: (timeStamp: ?number) => void;
  startJSSamplingProfiler: () => void;
  stopJSSamplingProfilerAndSaveToFile: (filePath: string) => void;
  setImageResponse(uri: string, imageResponse: ImageResponse): void;
  clearImage(uri: string): void;
  clearAllImages(): void;
}

export default TurboModuleRegistry.getEnforcing<Spec>(
  'NativeFantomCxx',
) as Spec;

File: private/react-native-fantom/tester/src/NativeFantom.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "NativeFantom.h"

#include <jsi/JSIDynamic.h>
#include <react/bridging/Bridging.h>
#include <react/renderer/components/modal/ModalHostViewShadowNode.h>
#include <react/renderer/components/scrollview/ScrollViewShadowNode.h>
#include <react/renderer/uimanager/UIManagerBinding.h>
#include <iostream>

#include "TesterAppDelegate.h"

#include <jsi/instrumentation.h>
#include "render/RenderFormatOptions.h"
#include "render/RenderOutput.h"

namespace facebook::react {

NativeFantom::NativeFantom(
    TesterAppDelegate& appDelegate,
    std::shared_ptr<CallInvoker> jsInvoker)
    : NativeFantomCxxSpec<NativeFantom>(std::move(jsInvoker)),
      appDelegate_(appDelegate) {}

SurfaceId NativeFantom::startSurface(
    jsi::Runtime& runtime,
    double viewportWidth,
    double viewportHeight,
    double devicePixelRatio,
    double viewportOffsetX,
    double viewportOffsetY) {
  SurfaceId surfaceId = nextSurfaceId_;
  nextSurfaceId_ += 10;
  appDelegate_.startSurface(
      runtime,
      static_cast<float>(viewportWidth),
      static_cast<float>(viewportHeight),
      surfaceId,
      static_cast<float>(devicePixelRatio),
      static_cast<float>(viewportOffsetX),
      static_cast<float>(viewportOffsetY));
  return surfaceId;
}

void NativeFantom::stopSurface(jsi::Runtime& /*runtime*/, SurfaceId surfaceId) {
  appDelegate_.stopSurface(surfaceId);
}

void NativeFantom::produceFramesForDuration(
    jsi::Runtime& /*runtime*/,
    double milliseconds) {
  appDelegate_.produceFramesForDuration(milliseconds);
}

void NativeFantom::flushMessageQueue(jsi::Runtime& /*runtime*/) {
  appDelegate_.flushMessageQueue();
}

void NativeFantom::flushEventQueue(jsi::Runtime& /*runtime*/) {
  appDelegate_.onRender();
}

void NativeFantom::validateEmptyMessageQueue(jsi::Runtime& /*runtime*/) {
  if (appDelegate_.hasPendingTasksInMessageQueue()) {
    throw std::runtime_error("MessageQueue is not empty");
  }
}

std::vector<std::string> NativeFantom::takeMountingManagerLogs(
    jsi::Runtime& /*runtime*/,
    SurfaceId surfaceId) {
  return appDelegate_.mountingManager_->takeMountingLogs(surfaceId);
}

std::string NativeFantom::getRenderedOutput(
    jsi::Runtime& /*runtime*/,
    SurfaceId surfaceId,
    NativeFantomGetRenderedOutputRenderFormatOptions options) {
  RenderFormatOptions formatOptions{
      options.includeRoot, options.includeLayoutMetrics};

  auto viewTree = appDelegate_.mountingManager_->getViewTree(surfaceId);
  return RenderOutput::render(viewTree, formatOptions);
}

void NativeFantom::reportTestSuiteResultsJSON(
    jsi::Runtime& /*runtime*/,
    std::string testSuiteResultsJSON) {
  std::cout << testSuiteResultsJSON << std::endl;
}

jsi::Object NativeFantom::getDirectManipulationProps(
    jsi::Runtime& runtime,
    const ShadowNode::Shared& shadowNode) {
  auto props = appDelegate_.mountingManager_->getViewDirectManipulationProps(
      shadowNode->getTag());
  return facebook::jsi::valueFromDynamic(runtime, props).asObject(runtime);
}

jsi::Object NativeFantom::getFabricUpdateProps(
    jsi::Runtime& runtime,
    const ShadowNode::Shared& shadowNode) {
  auto props = appDelegate_.mountingManager_->getViewFabricUpdateProps(
      shadowNode->getTag());
  return facebook::jsi::valueFromDynamic(runtime, props).asObject(runtime);
}

void NativeFantom::enqueueNativeEvent(
    jsi::Runtime& /*runtime*/,
    ShadowNode::Shared shadowNode,
    std::string type,
    const std::optional<folly::dynamic>& payload,
    std::optional<RawEvent::Category> category,
    std::optional<bool> isUnique) {
  if (isUnique.value_or(false)) {
    shadowNode->getEventEmitter()->dispatchUniqueEvent(
        std::move(type), payload.value_or(folly::dynamic::object()));
  } else {
    shadowNode->getEventEmitter()->dispatchEvent(
        std::move(type),
        payload.value_or(folly::dynamic::object()),
        category.value_or(RawEvent::Category::Unspecified));
  }
}

void NativeFantom::enqueueScrollEvent(
    jsi::Runtime& /*runtime*/,
    ShadowNode::Shared shadowNode,
    ScrollOptions options) {
  const auto* scrollViewShadowNode =
      dynamic_cast<const ScrollViewShadowNode*>(&*shadowNode);

  if (scrollViewShadowNode == nullptr) {
    throw std::runtime_error(
        "enqueueScrollEvent() can only be called on <ScrollView />");
  }

  auto point = Point{
      .x = options.x,
      .y = options.y,
  };

  auto scrollEvent = ScrollEvent();

  scrollEvent.contentOffset = point;
  scrollEvent.contentSize =
      scrollViewShadowNode->getStateData().getContentSize();
  scrollEvent.containerSize =
      scrollViewShadowNode->getLayoutMetrics().frame.size;
  scrollEvent.contentInset =
      scrollViewShadowNode->getConcreteProps().contentInset;
  scrollEvent.zoomScale = options.zoomScale.value_or(scrollEvent.zoomScale);

  scrollViewShadowNode->getConcreteEventEmitter().onScroll(scrollEvent);

  auto state =
      std::static_pointer_cast<const ScrollViewShadowNode::ConcreteState>(
          scrollViewShadowNode->getState());
  state->updateState(
      [point](const ScrollViewShadowNode::ConcreteState::Data& oldData)
          -> ScrollViewShadowNode::ConcreteState::SharedData {
        auto newData = oldData;
        newData.contentOffset = point;
        return std::make_shared<
            const ScrollViewShadowNode::ConcreteState::Data>(newData);
      });
}

void NativeFantom::enqueueModalSizeUpdate(
    jsi::Runtime& /*runtime*/,
    ShadowNode::Shared shadowNode,
    double width,
    double height) {
  const auto* modalHostViewShadowNode =
      dynamic_cast<const ModalHostViewShadowNode*>(&*shadowNode);

  if (modalHostViewShadowNode == nullptr) {
    throw std::runtime_error(
        "enqueueModalSizeUpdate() can only be called on <Modal />");
  }

  auto state =
      std::static_pointer_cast<const ModalHostViewShadowNode::ConcreteState>(
          modalHostViewShadowNode->getState());

  state->updateState(ModalHostViewState(
      {.width = static_cast<Float>(width),
       .height = static_cast<Float>(height)}));
}

jsi::Function NativeFantom::createShadowNodeReferenceCounter(
    jsi::Runtime& runtime,
    ShadowNode::Shared shadowNode) {
  auto weakShadowNode = std::weak_ptr<const ShadowNode>(shadowNode);

  return jsi::Function::createFromHostFunction(
      runtime,
      jsi::PropNameID::forAscii(runtime, "getReferenceCount"),
      0,
      [weakShadowNode](
          jsi::Runtime&, const jsi::Value&, const jsi::Value*, size_t)
          -> jsi::Value { return {(int)weakShadowNode.use_count()}; });
}

jsi::Function NativeFantom::createShadowNodeRevisionGetter(
    jsi::Runtime& runtime,
    ShadowNode::Shared shadowNode) {
#if RN_DEBUG_STRING_CONVERTIBLE
  auto weakShadowNode = std::weak_ptr<const ShadowNode>(shadowNode);

  return jsi::Function::createFromHostFunction(
      runtime,
      jsi::PropNameID::forAscii(runtime, "getRevision"),
      0,
      [weakShadowNode](
          jsi::Runtime& runtime, const jsi::Value&, const jsi::Value*, size_t)
          -> jsi::Value {
        if (auto strongShadowNode = weakShadowNode.lock()) {
          const auto& uiManager =
              UIManagerBinding::getBinding(runtime)->getUIManager();

          const auto& currentRevision =
              *uiManager.getNewestCloneOfShadowNode(*strongShadowNode);
          return currentRevision.revision_;
        } else {
          return jsi::Value::null();
        }
      });
#else
  // TODO(T225400348): Remove this when revision_ is available in optimised
  // builds.
  throw std::runtime_error(
      "createShadowNodeRevisionGetter() is only available in debug builds");
#endif
}

void NativeFantom::saveJSMemoryHeapSnapshot(
    jsi::Runtime& runtime,
    std::string filePath) {
  runtime.instrumentation().collectGarbage("heapsnapshot");
  runtime.instrumentation().createSnapshotToFile(filePath);
}

} // namespace facebook::react
